src/astarte_device.c

```diff
@@ -422,42 +422,83 @@ void astarte_device_destroy(astarte_device_handle_t device)
     free(device->encoded_hwid);
     free(device->credentials_secret);
     free(device->realm);
-    astarte_list_head_t *item = NULL;
-    astarte_list_head_t *tmp = NULL;
-    MUTABLE_LIST_FOR_EACH(item, tmp, &device->interfaces_list)
-    {
-        astarte_ptr_list_entry_t *entry = GET_LIST_ENTRY(item, astarte_ptr_list_entry_t, head);
-        free(entry);
-    }
+    astarte_linked_list_destroy(&device->introspection);
     free(device);
 }
 
 astarte_err_t astarte_device_add_interface(
     astarte_device_handle_t device, const astarte_interface_t *interface)
 {
+    astarte_err_t result = ASTARTE_OK;
     if (xSemaphoreTake(device->reinit_mutex, (TickType_t) 10) == pdFALSE) {
         ESP_LOGE(TAG, "Trying to add an interface to a device that is being reinitialized");
         return ASTARTE_ERR;
     }
 
     if (interface->major_version == 0 && interface->minor_version == 0) {
         ESP_LOGE(TAG, "Trying to add an interface with both major and minor version equal 0");
-        return ASTARTE_ERR_INVALID_INTERFACE_VERSION;
-    }
+        result = ASTARTE_ERR_INVALID_INTERFACE_VERSION;
+        goto end;
+    }
+
+    // Loop over any interface in introspection searching for an interface with the same name
+    size_t interface_name_len = strlen(interface->name);
+    astarte_linked_list_iterator_t list_iter;
+    astarte_err_t iter_err = astarte_linked_list_iterator_init(&device->introspection, &list_iter);
+    while (iter_err != ASTARTE_ERR_NOT_FOUND) {
+        astarte_interface_t *tmp_interface = astarte_linked_list_iterator_get_item(&list_iter);
+
+        // Check if the interface has the same name
+        size_t tmp_interface_name_len = strlen(tmp_interface->name);
+        size_t interfaces_min_len = (interface_name_len < tmp_interface_name_len)
+            ? interface_name_len
+            : tmp_interface_name_len;
+        if (strncmp(interface->name, tmp_interface->name, interfaces_min_len) == 0) {
+            ESP_LOGW(TAG, "Trying to add an interface already present in introspection");
+            // Check if ownership and type are the same
+            if ((interface->ownership != tmp_interface->ownership)
+                || (interface->type != tmp_interface->type)) {
+                ESP_LOGE(TAG, "Interface ownership/type conflicts with the one in introspection");
+                result = ASTARTE_ERR_CONFLICTING_INTERFACE;
+                goto end;
+            }
+            // Check if major versions align correctly
+            if (interface->major_version < tmp_interface->major_version) {
+                ESP_LOGE(TAG, "Interface with smaller major version than one in introspection");
+                result = ASTARTE_ERR_CONFLICTING_INTERFACE;
+                goto end;
+            }
+            if (interface->major_version == tmp_interface->major_version) {
+                // Check if minor versions aligns correctly
+                if (interface->minor_version < tmp_interface->minor_version) {
```

Comment:
You could avoid using 2 if statements and instead have only 1 of them by using the && operator